kratos/utilities/quaternion.h

```diff
@@ -94,9 +100,27 @@ namespace Kratos
 			}
 			return *this;
 		}
-                
-                
-#ifdef KRATOS_USE_AMATRIX   // This macro definition is for the migration period and to be removed afterward please do not use it 
+
+		const_reference	operator []	(size_t i) const
+		{
+			if (i == 0) return mX;
+			if (i == 1) return mY;
+			if (i == 2) return mZ;
+			if (i == 3) return mW;
+			KRATOS_ERROR << "Unknown component for a Quaternion. Component number "<<i<<"?";
+		}
+
+		reference operator [] (size_t i)
+		{
+			if (i == 0) return mX;
+			if (i == 1) return mY;
+			if (i == 2) return mZ;
+			if (i == 3) return mW;
+			KRATOS_ERROR << "Unknown component for a Quaternion. Component number "<<i<<"?";
+		}

```

Comment:
Performance-wise talking, I would change the internal doubles to `array_1d<double,4>` and here would only forward the operator.

Also you may expose it as an array in Python which may become useful.

Reply:
Hi @pooyan-dadvand , do you mean that we should construct a temporary vector always, like this?
```
array_1d<double,4> temp;
temp[0] = mX;
temp[1] = mY;
temp[2] = mZ;
temp[3] = mW;
return temp[i];
```

Reply:
IMO this code can be refactored and improved

Reply:
@loumalouomega and @pooyan-dadvand , first I would like to understand if you guys mean refactoring the method or refactoring the whole class and how it stores the member variables.

Reply:
> Hi @pooyan-dadvand , do you mean that we should construct a temporary vector always, like this?
> 
> ```
> array_1d<double,4> temp;
> temp[0] = mX;
> temp[1] = mY;
> temp[2] = mZ;
> temp[3] = mW;
> return temp[i];
> ```

I have not explained myself correctly. I would replace the `mX`, `mY`, `mZ`, `mW` with one `array_1d<double,4> mCoordinates` and then here I would directly return the `mCoordinates[i]` This also means that you should change every other place where those members are used with` mCoordinates[0]`, .... however optimizer would do its magic and other usage won't have overheads.

Reply:
I modified the code as suggested. The only doubt was in the constructors of the Quaternion.
Now the value of the components is given in the body of the method.
Also, the 0.0 value will only be valid if T is a double. I wonder if there is a generic way of calling the 0.

Reply:
It seems like we could use 
`mQuaternionValues[0] = T();`
I think this would initialize each component to zero, but I am not 100% sure.